cmd/gator/validate/validate.go

```diff
@@ -0,0 +1,200 @@
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var (
+	errLog *log.Logger
+	outLog *log.Logger
+)
+
+const (
+	examples = `  # Validate a manifest containing Kubernetes objects, Constraint Templates, and Constraints
+  gator validate --filename="manifest.yaml"
+
+  # Validate a directory
+  gator validate --filename="config-and-policies/"
+
+  # Use multiple inputs
+  gator validate --filename="manifest.yaml" --filename="templates-and-constraints/"
+
+  # Receive input from stdin
+  cat manifest.yaml | gator validate
+
+  # Output structured violations data
+  gator validate --filename="manifest.yaml" --json`
+)
+
+var Cmd = &cobra.Command{
+	Use:     "validate",
+	Short:   "validate resources against templates and constraints",
+	Example: examples,
+	Run:     run,
+}
+
+var (
+	flagFilenames []string
+	flagYAML      bool
+	flagJSON      bool
+)
+
+const (
+	flagNameFilename = "filename"
+	flagNameJSON     = "json"
+	flagNameYAML     = "yaml"
+)
+
+func init() {
+	Cmd.Flags().StringArrayVarP(&flagFilenames, flagNameFilename, "f", []string{}, "a file or directory containing kubernetes resources.  Can be specified multiple times.  Cannot be used in tandem with stdin.")
+	Cmd.Flags().BoolVar(&flagYAML, flagNameYAML, false, "print validation information as yaml.")
+	Cmd.Flags().BoolVar(&flagJSON, flagNameJSON, false, "print validation information as json.")
+
+	errLog = log.New(os.Stderr, "", 0)
+	outLog = log.New(os.Stdout, "", 0)
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var unstrucs []*unstructured.Unstructured
+
+	// check if stdin has data
+	stdinfo, err := os.Stdin.Stat()
+	if err != nil {
+		errLog.Fatalf("getting info for stdout: %s", err)
+	}
+
+	// using stdin in combination with flags is not supported
+	if stdinfo.Size() > 0 && len(flagFilenames) > 0 {
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	if stdinfo.Size() > 0 {
+		us, err := readYAMLSource(os.Stdin)
+		if err != nil {
+			errLog.Fatalf("reading from stdin: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	} else if len(flagFilenames) > 0 {
+		// normalize directories by listing their files
+		normalized, err := normalize(flagFilenames)
+		if err != nil {
+			errLog.Fatalf("normalizing: %s", err)
+		}
+
+		for _, filename := range normalized {
+			file, err := os.Open(filename)
+			if err != nil {
+				errLog.Fatalf("opening file %q: %s", filename, err)
+			}
+
+			us, err := readYAMLSource(bufio.NewReader(file))
+			if err != nil {
+				errLog.Fatalf("reading from file %q: %s", filename, err)
+			}
+			file.Close()
+
+			unstrucs = append(unstrucs, us...)
+		}
+	} else {
+		errLog.Fatalf("no input data: must include data via either stdin or the %q flag", flagNameFilename)
+	}
+
+	responses, err := validate.Validate(unstrucs)
+	if err != nil {
+		errLog.Fatalf("auditing objects: %v\n", err)
+	}
+
+	results := responses.Results()
+
+	if flagYAML && flagJSON {
```

Comment:
We don't need to worry about this if we use `--output=X`

Reply:
Added a TODO